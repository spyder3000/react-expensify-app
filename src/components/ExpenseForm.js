import React from 'react'; 
import moment from 'moment'; 
import { SingleDatePicker } from 'react-dates';  
import 'react-dates/lib/css/_datepicker.css'; 

//const date = new Date();    // from core JS
const now = moment();   // returns an instance of moment (e.g. current time); 
//console.log(now);   // on Chrome console, click _proto_ to see all functions available for this
console.log(now.format('YYYY-MM-DD')); 
console.log(now.format('MMM Do, YYYY')); 

export default class ExpenseForm extends React.Component {
    /* goal is to track changes for all of these fields which will then be used upon button submit */ 
    constructor(props) {
        super(props); 
        this.state = {
            description: props.expense ? props.expense.description : '', 
            note: props.expense ? props.expense.note : '',  
            amount: props.expense ? (props.expense.amount / 100).toString() : '',    // convert from number (cents) to string w/ decimal
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(), 
            calendarFocused: false, 
            error: ''            
        }
    }
    onDescriptionChange = (e) => {
        const description = e.target.value; 
        this.setState(() => ({ description })); 
    }
    onNoteChange = (e) => {
        const note = e.target.value; 
        this.setState(() => ({ note}));   // implicity return an object which sets note state to note variable
    }
    onAmountChange = (e) => {
        const amount = e.target.value; 
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {    // !amount allows user to clear value;  regex101.com to test regular expressions
            this.setState(() => ({ amount})); 
        }
    }
    onDateChange = (createdAt) => {
        if (createdAt) {        // prevents user from being able to delete date;  
            this.setState(() => ({createdAt}))
        }
    }; 
    onFocusChange = ({focused}) => {
        this.setState(() => ({calendarFocused: focused}))
    }; 

    onSubmit = ((e) => {
        e.preventDefault();  // to prevent full page refresh, handle w/ JS instead
        if (!this.state.description || !this.state.amount)  { // render error
            this.setState(() => ({ error: 'Please provide description and amount'}))
        }
        else {
            this.setState(() => ({ error: ''}))
            this.props.onSubmit({
                description: this.state.description, 
                amount: parseFloat(this.state.amount, 10) * 100,    // convert from string & from cents to dollar amt
                createdAt: this.state.createdAt.valueOf(),          // .valueOf() converts via momentjs.com Unix Timestamp documentation
                note: this.state.note
            })
        }
    })
    render() {
        return (
            <div>
                {this.state.error && <p>{this.state.error}</p>}
                <form onSubmit={this.onSubmit}>
                    <input 
                        type="text"
                        placeholder="Description"
                        autoFocus
                        value={this.state.description}
                        onChange={this.onDescriptionChange}
                    />
                    <input 
                        //type="number"
                        type="text"
                        placeholder="Amount"
                        value={this.state.amount} 
                        onChange={this.onAmountChange}
                    />
                    {/*  by default, SingleDatePicker only lets you choose current or future dates */}
                    <SingleDatePicker
                        date={this.state.createdAt} // momentPropTypes.momentObj or null
                        onDateChange={this.onDateChange} // PropTypes.func.isRequired
                        focused={this.state.calendarFocused} // PropTypes.bool
                        onFocusChange={this.onFocusChange} // PropTypes.func.isRequired
                        numberOfMonths = {1}    // default is 2
                        isOutsideRange = {() => false }  // override here lets you choose dates in the past 
                        //id="your_unique_id" // PropTypes.string.isRequired,
                    />
                    <textarea  
                        placeholder="Add a note for your expense (optional)" 
                        value={this.state.note}
                        onChange={this.onNoteChange}
                    >
                    </textarea>
                    <button>Add Expense</button>

                </form>
            </div>
        )
    }
}