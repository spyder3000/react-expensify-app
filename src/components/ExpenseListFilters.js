import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from '../actions/filters';  // import all action creators

// ch 110 -- replace stateless component w/ class-based component;  all props are this.props now   
export class ExpenseListFilters extends React.Component {
  state = {
    calendarFocused: null   
  };
  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);   // call both actions
    this.props.setEndDate(endDate);
  };
  onFocusChange = (calendarFocused) => {    // simple function to change calendarFocused to new value
    this.setState(() => ({ calendarFocused }));   // shorthand for calendarFocused: calendarFocused 
  }
  onTextChange = (e) => {
    this.props.setTextFilter(e.target.value);
  };
  onSortChange = (e) => {
    if (e.target.value === 'date') {
      this.props.sortByDate();
    } else if (e.target.value === 'amount') {
      this.props.sortByAmount();
    }
  };
  render() {
    return (  // returns JSX 
      <div>
        <input
          type="text"
          value={this.props.filters.text}
          onChange={this.onTextChange}
        />
        <select
          value={this.props.filters.sortBy}
          onChange={this.onSortChange}
        >
          <option value="date">Date</option>
          <option value="amount">Amount</option>
        </select>
        <DateRangePicker
          startDate={this.props.filters.startDate}
          endDate={this.props.filters.endDate}
          onDatesChange={this.onDatesChange}
          focusedInput={this.state.calendarFocused}
          onFocusChange={this.onFocusChange}
          showClearDates={true}   /* adds button to reset/clear dates */
          numberOfMonths={1}
          isOutsideRange={() => false}
        />
      </div>
    );
  }
};

const mapStateToProps = (state) => ({
  filters: state.filters
});

const mapDispatchToProps = (dispatch) => ({    // implicitly returns object;  same as => { return {... } }
  setTextFilter: (text) => dispatch(setTextFilter(text)),   // dispatches action 'setTextFilter' w/ param of 'text' 
  sortByDate: () => dispatch(sortByDate()),
  sortByAmount: () => dispatch(sortByAmount()),
  setStartDate: (startDate) => dispatch(setStartDate(startDate)),
  setEndDate: (endDate) => dispatch(setEndDate(endDate))
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);  // mapStateToProps to get us current data from state
