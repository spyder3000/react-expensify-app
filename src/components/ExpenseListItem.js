import React from 'react';   // all JSX is converted into React.createElement, so React here is needed, even for stateless components;
import { Link } from 'react-router-dom'; 
//import { removeExpense, editExpense } from '../actions/expenses'
//import { connect } from 'react-redux';  

// Export a stateless functional component 
//   description, amount, createdAt

const ExpenseListItem = ({dispatch,description, amount, createdAt, id}) => (      /* dispatch can be accessed via destructred var also */
    <div>
        <Link to={`/edit/${id}`}>
          <h3>{description}</h3>
        </Link> 

        <p>{amount} - {createdAt} </p>
        {/*<button onClick={() => {
          dispatch(removeExpense({id}))
        }}> Remove </button>  */}
    </div>
);

export default ExpenseListItem;   // moved Remove button to EditExpensePage.js, so no connect() needed;  
//export default connect()(ExpenseListItem);   // connect() because we don't want anything from state;  just want to connect to use dispatch()  

