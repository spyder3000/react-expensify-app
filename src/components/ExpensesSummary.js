import React from 'react'; 
import { connect } from 'react-redux';   // connect to connect to Redux Store
import numeral from 'numeral';  
import selectExpenses from '../selectors/expenses'; 
import selectTotal from '../selectors/expenses-total'; 
import selectExpensesTotal from '../selectors/expenses-total';

export const ExpensesSummary = ({ expenseCount, expensesTotal }) => {
    const expenseWord = expenseCount === 1 ? 'expense' : 'expenses' ; 
    const formattedExpensesTotal = numeral(expensesTotal / 100).format('$0,0.00'); 
    return (
        <div>
            <h1>Viewing {expenseCount} {expenseWord} totalling {formattedExpensesTotal} </h1>
        </div>
    )
}; 

// this is basically a function that returns a sorted & filtered array  
const mapStateToProps = (state) => {   // e.g. state.expenses, state.filters
    const visibleExpenses = selectExpenses(state.expenses, state.filters);  

    return { 
        expenseCount: visibleExpenses.length, 
        expensesTotal: selectExpensesTotal(visibleExpenses)
    };
}; 

// when you connect a component to the Redux store it's reactive, meaning changes will cause the page to re-render 
export default connect(mapStateToProps)(ExpensesSummary);    
