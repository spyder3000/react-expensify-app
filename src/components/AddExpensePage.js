import React from 'react';   // using ES6 syntax for React;  
import ExpenseForm from './ExpenseForm'; 
import { connect } from 'react-redux'; 
import { startAddExpense } from '../actions/expenses'; 

// ch 124 - convert to class-based component 
export class AddExpensePage extends React.Component {
    onSubmit= (expense) => {
        console.log(expense); 
       // props.dispatch(addExpense(expense));  // action creator that expects object w/ description, note, amount, createdAt
        this.props.startAddExpense(expense);  // replaces above line w/ this & mapDispatchToProps below
        this.props.history.push('/');   // switch to Dashboard page;  from Chrome, click 'Components' then 'AddExpensePage' to see history object
    }
    render() {
        return (
            <div>
            <h1>Add Expense</h1>
            <ExpenseForm onSubmit={this.onSubmit}
            /> 
        </div>
        )
    }
}

// goal is to return an object;  to more easily test, we want to replace props.dispatch(addExpense(expense)) above w/ props.onSubmit(expense)
const mapDispatchToProps = (dispatch) => ({     // implicitly returns object;  same as => { return { onsubmit: ... } }
    startAddExpense: (expense) => dispatch(startAddExpense(expense))
    }); 

export default connect(undefined, mapDispatchToProps)(AddExpensePage);   // () because we don't need anything from state;  this will give us access to props.dispatch above
