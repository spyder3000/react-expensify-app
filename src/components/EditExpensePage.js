import React from 'react';   // using ES6 syntax for React;  
import { connect } from 'react-redux'; 
import ExpenseForm from './ExpenseForm'; 
import { editExpense, startRemoveExpense } from '../actions/expenses'; 

export class EditExpensePage extends React.Component {
    onSubmit= (expense) => {
        console.log(expense); 
        // this.props.dispatch(editExpense(this.props.expense.id, expense));  // params - id & object w/ description, note, amount, createdAt
        this.props.editExpense(this.props.expense.id, expense);  // replaces above line by removing dispatch which is in mapDispatchToProps below
        this.props.history.push('/');   // switch to Dashboard page;  from Chrome, click 'Components' then 'AddExpensePage' to see history object
    }
    onRemove = () => {
        // this.props.dispatch(startRemoveExpense({id: this.props.expense.id}));
        this.props.startRemoveExpense({ id: this.props.expense.id });   
        this.props.history.push('/'); 
    }
    render() {
        return (
            <div>
                <ExpenseForm 
                    expense={this.props.expense}  // expense prop to hold existing expense object;  from mapStateToProps() below
                    onSubmit={this.onSubmit}    
                />
                <button onClick={this.onRemove}>Remove</button>
            </div>
        );
    }
}; 


const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find((expense) => {
            return expense.id === props.match.params.id;   // props.match.params.id comes from ExpenseListItem.js <Link>
        })
    }
}

// goal is to return an object;  to more easily test, we want to replace props.dispatch(editExpense(expense)) above w/ props.onSubmit(expense);  
//  2nd param is Own props -- it's there if needed, but not needed here 
const mapDispatchToProps = (dispatch, props) => ({     // implicitly returns object;  same as => { return { onsubmit: ... } }
    editExpense: (id, expense) => dispatch(editExpense(id, expense)), 
    startRemoveExpense: (data) => dispatch(startRemoveExpense(data))
}); 

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage); // mapStateToProps to get us current data from state
