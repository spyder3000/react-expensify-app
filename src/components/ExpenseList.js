import React from 'react'; 
import { connect } from 'react-redux';   // connect to connect to Redux Store
import ExpenseListItem from './ExpenseListItem';  
import selectExpenses from '../selectors/expenses'; 

export const ExpenseList = (props) => (     // set to export to test in ExpenseList.test.js 
    <div>
        { props.expenses.length === 0 ? (
            <p>No expenses</p>
        ) : (
            props.expenses.map((expense) => {
                return <ExpenseListItem key={expense.id} {...expense} />   /* ...expense allows for destructuring in ExpenseListItem */
            })
        )
        }
    </div>
); 

// replaces logic below (this one's more standard);  this is basically a function that returns a sorted & filtered array  
const mapStateToProps = (state) => {   // e.g. state.expenses, state.filters
    return {
        expenses: selectExpenses(state.expenses, state.filters)   // params for default fn in selectors/expenses
        // expenses: state.expenses, 
        // filters: state.filters
    }; 
}; 

// when you connect a component to the Redux store it's reactive, meaning changes will cause the page to re-render 
export default connect(mapStateToProps)(ExpenseList);    

// const ConnectedExpenseList = connect((state) => {   // e.g. state.expenses, state.filters
//     return {
//         expenses: state.expenses
//     }; 
// })(ExpenseList);  // connect() returns a function which we need to call w/ the component ExpenseList;  
// export default ConnectedExpenseList; 