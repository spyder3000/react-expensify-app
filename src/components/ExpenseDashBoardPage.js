import React from 'react';   // using ES6 syntax for React;  
import ExpenseList from './ExpenseList';  
import ExpenseListFilters from './ExpenseListFilters';  
import ExpensesSummary from './ExpensesSummary';  

const ExpenseDashBoardPage = () => (
    <div>
        <ExpensesSummary /> 
        <ExpenseListFilters /> 
        <ExpenseList />
    </div>
);

export default ExpenseDashBoardPage; 