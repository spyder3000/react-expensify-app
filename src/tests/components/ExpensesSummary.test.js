import React from 'react'; 
import { shallow } from 'enzyme'; 
import { ExpensesSummary }  from '../../components/ExpensesSummary'; 

// test here is to test the const ExpenseListItem component in ExpenseListItem.js  
test('Should render ExpensesSummary correctly for one expense', () => {
    const wrapper =  shallow(<ExpensesSummary expenseCount={1} expensesTotal={3333} />);   
    expect(wrapper).toMatchSnapshot(); 
})

test('Should render ExpensesSummary correctly for multiple expenses', () => {
    const wrapper =  shallow(<ExpensesSummary expenseCount={8} expensesTotal={49033} />); 
    expect(wrapper).toMatchSnapshot(); 
})