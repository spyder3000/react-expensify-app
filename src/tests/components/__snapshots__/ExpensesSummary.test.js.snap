// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render ExpensesSummary correctly for multiple expenses 1`] = `
<div>
  <h1>
    Viewing 
    8
     
    expenses
     totalling 
    $490.33
     
  </h1>
</div>
`;

exports[`Should render ExpensesSummary correctly for one expense 1`] = `
<div>
  <h1>
    Viewing 
    1
     
    expense
     totalling 
    $33.33
     
  </h1>
</div>
`;
